// @vuetify/v0/dist/data-table/index.mjs
import "../globals-DCznFHqN.mjs";
import "../createTrinity-C_c8eedz.mjs";
import "../useLogger-C3crtWEd.mjs";
import "../pino-B7tJ3m18.mjs";
import "../v0-2IcFDTW5.mjs";
import "../adapters-Bv5aRv3l.mjs";
import "../createTokens-CpapxkYX.mjs";
import "../createSelection-BsHsL_qW.mjs";
import "../createSingle-BBwE5j_k.mjs";
import "../createGroup-BqN4Ve83.mjs";
import "../toArray-VdROPpy-.mjs";
import "../v0-CB-GCB-s.mjs";
import "../adapters-BjeDmKc6.mjs";
import "../useLocale-D3btFSZv.mjs";
import "../createPagination-CeGMgKCx.mjs";
import { t as DataTableAdapter } from "../adapter-DXRKF27V.mjs";
import { t as ClientAdapter } from "../v0-D87sooe0.mjs";
import { t as ServerAdapter } from "../server-CYYGvVps.mjs";
import { t as VirtualAdapter } from "../virtual-D75KhU6Q.mjs";
import "../adapters-uJLWGQJl.mjs";
import { n as createDataTableContext, r as useDataTable, t as createDataTable } from "../createDataTable-DHxv--aj.mjs";
export { ClientAdapter, DataTableAdapter, ServerAdapter, VirtualAdapter, createDataTable, createDataTableContext, useDataTable };
