// @vuetify/v0/dist/adapter-DXRKF27V.mjs
import { isNaN, isNullOrUndefined, isNumber, isObject, isString } from "./utilities/index.mjs";
import { i as useContext, n as createContext, t as createTrinity } from "./createTrinity-C_c8eedz.mjs";
import { t as toArray } from "./toArray-VdROPpy-.mjs";
import { computed, shallowRef, toRef, toValue } from "vue";
//#region src/composables/createFilter/index.ts
/**
* @module useFilter
*
* @remarks
* Reactive array filtering composable with multiple filter modes.
*
* Key features:
* - Four filter modes: some, every, union, intersection
* - Case-insensitive filtering
* - Custom filter functions
* - Reactive updates
* - Context-based DI support
* - Perfect for search, multi-criteria filtering
*
* Filters arrays based on query strings with configurable matching strategies.
*/
function defaultFilter(query, item, keys, mode = "some") {
	const queries = toArray(query).map((q) => String(q).toLowerCase());
	const stringValues = (/* @__PURE__ */ isObject(item) ? keys?.length ? keys.map((k) => item[k]) : Object.values(item) : [item]).map((v) => String(v).toLowerCase());
	if (mode === "some") return queries.some((q) => stringValues.some((val) => val.includes(q)));
	if (mode === "every") return queries.every((q) => stringValues.every((val) => val.includes(q)));
	const joined = stringValues.join("\0");
	if (mode === "union") return queries.some((q) => joined.includes(q));
	if (mode === "intersection") return queries.every((q) => joined.includes(q));
	return false;
}
/**
* Creates a filter context with pre-configured options.
*
* @param options The filter options
* @template Z The type of the items
* @template E The type of the filter context
* @returns A filter context
*
* @see https://0.vuetifyjs.com/composables/utilities/use-filter
*
* @example
* ```ts
* import { createFilter } from '@vuetify/v0'
*
* const filter = createFilter({
*   mode: 'intersection',
*   keys: ['name', 'email'],
* })
*
* const { items } = filter.apply(query, users)
* ```
*/
function createFilter(options = {}) {
	const { customFilter, keys, mode = "some" } = options;
	const filterFunction = customFilter ?? ((q, i) => defaultFilter(q, i, keys, mode));
	const query = shallowRef("");
	function apply(_query, items) {
		return { items: computed(() => {
			const q = toValue(_query);
			query.value = q;
			const queries = toArray(q).filter((q) => String(q).trim());
			if (queries.length === 0) return toValue(items);
			const queryParam = queries.length === 1 ? queries[0] : queries;
			return toValue(items).filter((item) => filterFunction(queryParam, item));
		}) };
	}
	return {
		mode,
		keys,
		customFilter,
		query,
		apply
	};
}
/**
* Creates a filter context with dependency injection support.
*
* @param options The filter context options
* @template Z The type of the items
* @template E The type of the filter context
* @returns A trinity tuple: [useContext, provideContext, defaultContext]
*
* @see https://0.vuetifyjs.com/composables/utilities/use-filter
*
* @example
* ```ts
* import { createFilterContext } from '@vuetify/v0'
*
* export const [useSearchFilter, provideSearchFilter, searchFilter] = createFilterContext({
*   namespace: 'app:search',
*   mode: 'union',
*   keys: ['title', 'description'],
* })
*
* // In parent component
* provideSearchFilter()
*
* // In child component
* const filter = useSearchFilter()
* const { items } = filter.apply(query, products)
* ```
*/
function createFilterContext(_options = {}) {
	const { namespace = "v0:filter", ...options } = _options;
	const [useFilterContext, _provideFilterContext] = createContext(namespace);
	const context = createFilter(options);
	function provideFilterContext(_context = context, app) {
		return _provideFilterContext(_context, app);
	}
	return createTrinity(useFilterContext, provideFilterContext, context);
}
/**
* Returns the current filter context from dependency injection.
*
* @param namespace The namespace for the filter context. Defaults to `'v0:filter'`.
* @template Z The type of the items.
* @template E The type of the filter context.
* @returns The current filter context.
*
* @see https://0.vuetifyjs.com/composables/utilities/use-filter
*
* @example
* ```vue
* <script setup lang="ts">
*   import { useFilter } from '@vuetify/v0'
*
*   const filter = useFilter()
*   const { items } = filter.apply(query, products)
* <\/script>
* ```
*/
function useFilter(namespace = "v0:filter") {
	return useContext(namespace);
}
//#endregion
//#region src/composables/createDataTable/adapters/adapter.ts
/**
* @module createDataTable/adapters
*
* @remarks
* Defines the adapter interface and abstract base class for data table
* pipeline strategies. Adapters control how raw items are filtered, sorted,
* and sliced into visible items.
*
* The abstract DataTableAdapter provides shared filter and sort pipelines
* used by client-side adapters. Server adapters that delegate pipeline
* operations externally can implement DataTableAdapterInterface directly.
*/
function getNestedValue(obj, key) {
	const keys = key.split(".");
	let result = obj;
	for (const k of keys) {
		if (/* @__PURE__ */ isNullOrUndefined(result)) return void 0;
		result = result[k];
	}
	return result;
}
function compareValues(a, b, locale) {
	if (a === b) return 0;
	const aNil = /* @__PURE__ */ isNullOrUndefined(a);
	const bNil = /* @__PURE__ */ isNullOrUndefined(b);
	if (aNil && bNil) return 0;
	if (aNil) return 1;
	if (bNil) return -1;
	if (/* @__PURE__ */ isString(a) && /* @__PURE__ */ isString(b)) return a.localeCompare(b, locale);
	if (/* @__PURE__ */ isNumber(a) && /* @__PURE__ */ isNumber(b)) {
		if (/* @__PURE__ */ isNaN(a)) return 1;
		if (/* @__PURE__ */ isNaN(b)) return -1;
		return a - b;
	}
	return String(a).localeCompare(String(b), locale);
}
var DataTableAdapter = class {
	/** Create the filter pipeline stage */
	filter(context) {
		const filter = createFilter(Object.keys(context.customColumnFilters).length > 0 && !context.filterOptions.customFilter ? {
			...context.filterOptions,
			keys: [...context.filterableKeys],
			customFilter: (query, item) => {
				if (!/* @__PURE__ */ isObject(item)) return false;
				const q = String(Array.isArray(query) ? query[0] : query).toLowerCase();
				if (!q) return true;
				const obj = item;
				for (const key of context.filterableKeys) {
					const customFn = context.customColumnFilters[key];
					if (customFn) {
						if (customFn(obj[key], q)) return true;
					} else {
						const val = obj[key];
						if (!/* @__PURE__ */ isNullOrUndefined(val) && String(val).toLowerCase().includes(q)) return true;
					}
				}
				return false;
			}
		} : {
			...context.filterOptions,
			keys: [...context.filterableKeys]
		});
		const allItems = toRef(() => toValue(context.items));
		const { items: filteredItems } = filter.apply(context.search, allItems);
		return {
			allItems,
			filteredItems
		};
	}
	/** Create the sort pipeline stage */
	sort(filteredItems, sortBy, locale, customSorts) {
		return computed(() => {
			const entries = sortBy.value;
			if (entries.length === 0) return filteredItems.value;
			const loc = locale?.value;
			return filteredItems.value.toSorted((a, b) => {
				for (const { key, direction } of entries) {
					const aVal = getNestedValue(a, key);
					const bVal = getNestedValue(b, key);
					const customSort = customSorts?.[key];
					const cmp = customSort ? customSort(aVal, bVal) : compareValues(aVal, bVal, loc);
					if (cmp !== 0) return direction === "desc" ? -cmp : cmp;
				}
				return 0;
			});
		});
	}
};
//#endregion
export { useFilter as i, createFilter as n, createFilterContext as r, DataTableAdapter as t };
