// @vuetify/v0/dist/createTrinity-C_c8eedz.mjs
import { isObject, isString, isSymbol, isUndefined } from "./utilities/index.mjs";
import { inject, provide } from "vue";
//#region src/composables/createContext/index.ts
/**
* @module createContext
*
* @see https://0.vuetifyjs.com/composables/foundation/create-context
*
* @remarks
* Factory for creating type-safe Vue dependency injection contexts.
*
* Provides a wrapper around Vue's provide/inject that throws errors when context is not found,
* eliminating silent failures and improving developer experience. Supports both app-level and
* component-level provision.
*
* Supports two modes:
* - **Static key**: `createContext('my-key')` - key is fixed at creation time
* - **Dynamic key**: `createContext()` or `createContext({ suffix: 'item' })` - key provided at runtime
*/
/**
* Injects a context provided by an ancestor component.
*
* @param key The key of the context to inject.
* @param defaultValue Optional default value if context is not found.
* @template Z The type of the context.
* @returns The injected context.
* @throws An error if the context is not found and no default is provided.
*
* @see https://vuejs.org/api/composition-api-dependency-injection.html#inject
* @see https://0.vuetifyjs.com/composables/foundation/create-context#use-context
*
* @example
* ```ts
* // Without default value
* const context = useContext<MyContext>('my-context')
*
* // With default value
* const context = useContext<MyContext>('my-context', defaultContext)
* ```
*/
function useContext(key, defaultValue) {
	const context = inject(key, defaultValue);
	if (/* @__PURE__ */ isUndefined(context)) throw new Error(`Context "${String(key)}" not found. Ensure it's provided by an ancestor.`);
	return context;
}
/**
* Provides a context to all descendant components.
*
* @param key The key of the context to provide.
* @param context The context to provide.
* @param app Optional Vue app instance to provide the context at app level instead of component level.
* @template Z The type of the context.
* @returns The provided context.
*
* @remarks
* When `app` parameter is provided, the context is made available to all components in the app.
* When omitted, the context is provided at the current component level and available to descendants only.
*
* @see https://vuejs.org/api/composition-api-dependency-injection.html#provide
* @see https://0.vuetifyjs.com/composables/foundation/create-context#provide-context
*
* @example
* ```ts
* // Component-level provision
* provideContext<MyContext>('my-context', context)
*
* // App-level provision (typically used in plugins)
* const app = createApp()
* provideContext<MyContext>('my-context', context, app)
* ```
*/
function provideContext(key, context, app) {
	if (app) app.provide(key, context);
	else provide(key, context);
	return context;
}
function createContext(keyOrOptions, defaultValue) {
	if (/* @__PURE__ */ isString(keyOrOptions) || /* @__PURE__ */ isSymbol(keyOrOptions)) {
		const _key = keyOrOptions;
		function _provideContext(context, app) {
			return provideContext(_key, context, app);
		}
		function _useContext() {
			return useContext(_key, defaultValue);
		}
		return [_useContext, _provideContext];
	}
	const suffix = /* @__PURE__ */ isObject(keyOrOptions) ? keyOrOptions.suffix : void 0;
	function _provideContext(key, context, app) {
		return provideContext(suffix ? `${key}:${suffix}` : key, context, app);
	}
	function _useContext(key, defaultValue) {
		return useContext(suffix ? `${key}:${suffix}` : key, defaultValue);
	}
	return [_useContext, _provideContext];
}
//#endregion
//#region src/composables/createTrinity/index.ts
/**
* Creates a new trinity for a context composable and its provider.
*
* @param useContext The function that retrieves/uses the context (typically named `useContext`).
* @param provideContext The function that provides the context to descendants.
* @param context The default context instance to use when no custom context is provided.
* @template Z The type of the context.
* @returns A readonly tuple containing: [useContext function, provideContext wrapper function, default context instance].
*
* @remarks The trinity pattern is a foundational pattern used throughout the codebase for creating reusable context systems. It provides three related elements:
*
* 1. A function to retrieve/use the context
* 2. A function to provide the context (with default value support)
* 3. The default context instance
*
* The returned tuple is readonly (using `as const`) to ensure proper type inference.
*
* @see https://0.vuetifyjs.com/composables/foundation/create-trinity#create-trinity
*
* @example
* ```ts
* interface MyContext {
*   foo: string
*   bar: number
* }
*
* export function createMyFeature<E extends MyContext = MyContext>() {
*   const [useContext, _provideContext] = createContext<E>('my-context')
*
*   const context = { foo: 'hello', bar: 42 }
*
*   function provideContext (_context: E = context, app?: App): E {
*     return _provideContext(_context, app)
*   }
*
*   return createTrinity<E>(useContext, provideContext, context)
* }
* ```
*/
function createTrinity(useContext, provideContext, context) {
	return [
		useContext,
		(_context = context, app) => provideContext(_context, app),
		context
	];
}
//#endregion
export { useContext as i, createContext as n, provideContext as r, createTrinity as t };
