// @vuetify/v0/dist/createPagination-CeGMgKCx.mjs
import { isNaN, range } from "./utilities/index.mjs";
import { i as useContext, n as createContext, t as createTrinity } from "./createTrinity-C_c8eedz.mjs";
import { computed, isRef, shallowRef, toRef, toValue } from "vue";
//#region src/composables/createPagination/index.ts
/**
* @module usePagination
*
* @remarks
* Lightweight pagination composable for navigating through pages.
*
* Key features:
* - No registry overhead - just a bounded integer
* - Direct ref support for v-model compatibility
* - Navigation methods: next, prev, first, last
* - Computed visible items with ellipsis
* - Trinity pattern for dependency injection
*
* Unlike registry-based composables, pagination tracks a single number
* within a range, making it efficient for large page counts.
*/
/**
* Creates a pagination instance.
*
* @param options The options for the pagination instance.
* @returns A pagination context with navigation methods.
*
* @example
* ```ts
* import { createPagination } from '@vuetify/v0'
*
* // Basic usage
* const pagination = createPagination({ size: 100 })
* pagination.next()
* pagination.items.value // [{ type: 'page', value: 1 }, { type: 'page', value: 2 }, ...]
*
* // With v-model (pass a ref)
* const page = ref(1)
* const pagination = createPagination({ page, size: 100 })
* // Mutating pagination.page or the passed ref syncs both
* ```
*/
function createPagination(_options = {}) {
	const { page: _page = 1, itemsPerPage: _itemsPerPage = 10, size: _size = 0, visible: _visible = 7, ellipsis = "..." } = _options;
	const _raw = isRef(_page) ? _page : shallowRef(_page);
	const pages = computed(() => {
		const size = toValue(_size);
		const perPage = toValue(_itemsPerPage);
		if (size <= 0 || /* @__PURE__ */ isNaN(size)) return 0;
		return Math.ceil(size / perPage);
	});
	const page = computed({
		get: () => {
			const p = pages.value;
			return p > 0 && _raw.value > p ? p : _raw.value;
		},
		set: (v) => {
			_raw.value = v;
		}
	});
	function first() {
		page.value = 1;
	}
	function last() {
		page.value = Math.max(1, pages.value);
	}
	function next() {
		if (page.value < pages.value) page.value++;
	}
	function prev() {
		if (page.value > 1) page.value--;
	}
	function select(value) {
		if (value < 1) page.value = 1;
		else if (value > pages.value) page.value = Math.max(1, pages.value);
		else page.value = value;
	}
	const isFirst = toRef(() => page.value <= 1);
	const isLast = toRef(() => page.value >= pages.value);
	const pageStart = toRef(() => (page.value - 1) * toValue(_itemsPerPage));
	const pageStop = toRef(() => Math.min(pageStart.value + toValue(_itemsPerPage), toValue(_size)));
	function toPage(value) {
		return {
			type: "page",
			value
		};
	}
	function toEllipsis() {
		return ellipsis === false ? false : {
			type: "ellipsis",
			value: ellipsis
		};
	}
	function filter(array) {
		return array.filter((item) => item !== false);
	}
	return {
		page,
		ellipsis,
		items: computed(() => {
			const pageCount = pages.value;
			const visible = toValue(_visible);
			const current = page.value;
			if (pageCount <= 0 || /* @__PURE__ */ isNaN(pageCount) || pageCount > Number.MAX_SAFE_INTEGER) return [];
			if (visible <= 0) return [];
			if (visible <= 2) return [toPage(current)];
			if (pageCount <= visible) return (/* @__PURE__ */ range(pageCount, 1)).map(toPage);
			if (visible === 3) {
				const mid = current <= 1 ? 2 : current >= pageCount ? pageCount - 1 : current;
				return [
					toPage(1),
					toPage(mid),
					toPage(pageCount)
				];
			}
			const boundary = visible - 2;
			const middle = visible - 4;
			if (middle <= 0) {
				if (current <= boundary) return filter([
					...(/* @__PURE__ */ range(boundary, 1)).map(toPage),
					toEllipsis(),
					toPage(pageCount)
				]);
				if (current > pageCount - boundary) return filter([
					toPage(1),
					toEllipsis(),
					...(/* @__PURE__ */ range(boundary, pageCount - boundary + 1)).map(toPage)
				]);
				return current <= Math.ceil(pageCount / 2) ? filter([
					toPage(1),
					toPage(current),
					toEllipsis(),
					toPage(pageCount)
				]) : filter([
					toPage(1),
					toEllipsis(),
					toPage(current),
					toPage(pageCount)
				]);
			}
			const leftThreshold = boundary - 1;
			const rightThreshold = pageCount - boundary + 2;
			if (current <= leftThreshold) return filter([
				...(/* @__PURE__ */ range(boundary, 1)).map(toPage),
				toEllipsis(),
				toPage(pageCount)
			]);
			else if (current >= rightThreshold) return filter([
				toPage(1),
				toEllipsis(),
				...(/* @__PURE__ */ range(boundary, pageCount - boundary + 1)).map(toPage)
			]);
			else {
				const start = current - Math.floor(middle / 2);
				return filter([
					toPage(1),
					toEllipsis(),
					...(/* @__PURE__ */ range(middle, start)).map(toPage),
					toEllipsis(),
					toPage(pageCount)
				]);
			}
		}),
		pageStart,
		pageStop,
		isFirst,
		isLast,
		first,
		last,
		next,
		prev,
		select,
		get itemsPerPage() {
			return toValue(_itemsPerPage);
		},
		get size() {
			return toValue(_size);
		},
		get pages() {
			return pages.value;
		}
	};
}
/**
* Creates a pagination context for dependency injection.
*
* @param options The options including namespace.
* @returns A trinity: [usePagination, providePagination, defaultContext]
*
* @example
* ```ts
* // With default namespace 'v0:pagination'
* const [usePagination, providePaginationContext] = createPaginationContext({ size: 50 })
*
* // Or with custom namespace
* const [usePagination, providePaginationContext] = createPaginationContext({
*   namespace: 'my-pagination',
*   size: 50,
* })
*
* // Parent component
* providePaginationContext()
*
* // Child component
* const pagination = usePagination()
* pagination.next()
* ```
*/
function createPaginationContext(_options = {}) {
	const { namespace = "v0:pagination", ...options } = _options;
	const [usePaginationContext, _providePaginationContext] = createContext(namespace);
	const context = createPagination(options);
	function providePaginationContext(_context = context, app) {
		return _providePaginationContext(_context, app);
	}
	return createTrinity(usePaginationContext, providePaginationContext, context);
}
/**
* Returns the current pagination instance from context.
*
* @param namespace The namespace. @default 'v0:pagination'
* @returns The pagination context.
*
* @example
* ```vue
* <script setup lang="ts">
*  import { usePagination } from '@vuetify/v0'
*
*  const pagination = usePagination()
* <\/script>
*
* <template>
*   <button @click="pagination.prev()" :disabled="pagination.isFirst.value">Prev</button>
*   <button @click="pagination.next()" :disabled="pagination.isLast.value">Next</button>
* </template>
* ```
*/
function usePagination(namespace = "v0:pagination") {
	return useContext(namespace);
}
//#endregion
export { createPaginationContext as n, usePagination as r, createPagination as t };
