// @vuetify/v0/dist/flagsmith-CrWUeIhT.mjs
//#region src/composables/useFeatures/adapters/flagsmith/index.ts
var FlagsmithFeatureAdapter = class {
	client;
	options;
	constructor(client, options) {
		this.client = client;
		this.options = options;
	}
	setup(onUpdate) {
		const updateFlags = () => {
			const flags = this.client.getAllFlags();
			const adapterFlags = {};
			if (flags) for (const [key, flag] of Object.entries(flags)) {
				const isEnabled = flag.enabled;
				const variation = flag.value;
				adapterFlags[key] = variation !== null && variation !== void 0 ? {
					$value: isEnabled,
					$variation: variation
				} : isEnabled;
			}
			return adapterFlags;
		};
		this.client.init({
			...this.options,
			onChange: (oldFlags, params, loadingState) => {
				onUpdate(updateFlags());
				this.options.onChange?.(oldFlags, params, loadingState);
			}
		});
		this.disposeFn = () => this.client.stopListening();
		return updateFlags();
	}
	dispose() {
		this.disposeFn();
	}
	disposeFn = () => {};
};
//#endregion
export { FlagsmithFeatureAdapter as t };
