// @vuetify/v0/dist/htmlElements-DcYtv0hg.mjs
/**
* Set of HTML elements that are self-closing (void elements)
* These elements cannot have children and don't need closing tags
*/
const SELF_CLOSING_TAGS = new Set([
	"area",
	"base",
	"br",
	"col",
	"embed",
	"hr",
	"img",
	"input",
	"link",
	"meta",
	"source",
	"track",
	"wbr"
]);
/**
* Common HTML element types for polymorphic components
*/
const COMMON_ELEMENTS = {
	DIV: "div",
	SPAN: "span",
	SECTION: "section",
	ARTICLE: "article",
	ASIDE: "aside",
	HEADER: "header",
	FOOTER: "footer",
	MAIN: "main",
	NAV: "nav",
	P: "p",
	H1: "h1",
	H2: "h2",
	H3: "h3",
	H4: "h4",
	H5: "h5",
	H6: "h6",
	BUTTON: "button",
	A: "a",
	INPUT: "input",
	TEXTAREA: "textarea",
	SELECT: "select",
	LABEL: "label",
	UL: "ul",
	OL: "ol",
	LI: "li",
	DL: "dl",
	DT: "dt",
	DD: "dd",
	IMG: "img",
	VIDEO: "video",
	AUDIO: "audio",
	CANVAS: "canvas",
	SVG: "svg",
	TABLE: "table",
	THEAD: "thead",
	TBODY: "tbody",
	TFOOT: "tfoot",
	TR: "tr",
	TH: "th",
	TD: "td",
	FORM: "form",
	FIELDSET: "fieldset",
	LEGEND: "legend"
};
/**
* Check if an element is self-closing
*/
function isSelfClosingTag(tag) {
	return SELF_CLOSING_TAGS.has(tag.toLowerCase());
}
//#endregion
export { SELF_CLOSING_TAGS as n, isSelfClosingTag as r, COMMON_ELEMENTS as t };
