// @vuetify/v0/dist/pino-B7tJ3m18.mjs
import { isObject } from "./utilities/index.mjs";
//#region src/composables/useLogger/adapters/pino.ts
/**
* Pino logger adapter implementation
*
* This adapter integrates with the Pino logging library,
* providing high-performance structured logging optimized
* for Node.js applications with minimal overhead.
*/
var PinoLoggerAdapter = class {
	pino;
	constructor(pinoInstance) {
		if (!pinoInstance) throw new Error("Pino instance is required for PinoLoggerAdapter");
		this.pino = pinoInstance;
	}
	debug(message, ...args) {
		this.pino.debug(this.format(message, ...args));
	}
	info(message, ...args) {
		this.pino.info(this.format(message, ...args));
	}
	warn(message, ...args) {
		this.pino.warn(this.format(message, ...args));
	}
	error(message, ...args) {
		this.pino.error(this.format(message, ...args));
	}
	trace(message, ...args) {
		this.pino.trace(this.format(message, ...args));
	}
	fatal(message, ...args) {
		this.pino.fatal(this.format(message, ...args));
	}
	format(message, ...args) {
		if (args.length === 0) return { msg: message };
		if (args.length === 1 && /* @__PURE__ */ isObject(args[0])) return {
			...args[0],
			msg: message
		};
		return {
			msg: message,
			args
		};
	}
};
//#endregion
export { PinoLoggerAdapter as t };
