// @vuetify/v0/dist/posthog-BugD9w3g.mjs
import { t as IN_BROWSER } from "./globals-DCznFHqN.mjs";
//#region src/composables/useFeatures/adapters/posthog/index.ts
/**
* @module PostHogFeatureAdapter
*
* @remarks
* PostHog adapter for feature flags.
*/
var PostHogFeatureAdapter = class {
	constructor(client) {
		this.client = client;
	}
	setup(onUpdate) {
		if (!IN_BROWSER) return {};
		const collect = () => {
			const flags = {};
			const activeFlags = this.client.featureFlags.getFlags();
			if (!activeFlags) return flags;
			for (const key of activeFlags) {
				const isEnabled = this.client.isFeatureEnabled(key) ?? false;
				const payload = this.client.getFeatureFlagPayload(key);
				if (payload !== void 0 && payload !== null) flags[key] = {
					$value: isEnabled,
					$variation: payload
				};
				else {
					const variant = this.client.getFeatureFlag(key);
					flags[key] = variant !== true && variant !== false && variant !== void 0 && variant !== null ? {
						$value: true,
						$variation: variant
					} : isEnabled;
				}
			}
			return flags;
		};
		this.disposeFn = this.client.onFeatureFlags(() => onUpdate(collect()));
		return collect();
	}
	dispose() {
		this.disposeFn();
	}
	disposeFn = () => {};
};
//#endregion
export { PostHogFeatureAdapter as t };
