// @vuetify/v0/dist/useFeatures-CkIfpj7l.mjs
import { isArray, isBoolean, isFunction, isObject } from "./utilities/index.mjs";
import { o as createPluginContext } from "./useLogger-C3crtWEd.mjs";
import { t as createTokens } from "./createTokens-CpapxkYX.mjs";
import { t as createGroup } from "./createGroup-BqN4Ve83.mjs";
//#region src/composables/useFeatures/adapters/generic/index.ts
var FeaturesAdapter = class {};
//#endregion
//#region src/composables/useFeatures/index.ts
/**
* @module useFeatures
*
* @see https://0.vuetifyjs.com/composables/plugins/use-features
*
* @remarks
* Feature flag system with boolean and token-based features.
*
* Key features:
* - Boolean features (true/false activation)
* - Token features with $variation support
* - Auto-selection of enabled features
* - Multi-select support for feature combinations
* - Perfect for A/B testing, progressive rollout, feature toggles
* - Adapter pattern for external feature flag services
*
* Inheritance chain: useRegistry → createSelection → createGroup → createFeatures
* Integrates with useTokens for token-based features.
*/
/**
* Creates a new features instance.
*
* @param options The options for the features instance.
* @template Z The type of the feature ticket.
* @template E The type of the feature context.
* @returns A new features instance.
*
* @see https://0.vuetifyjs.com/composables/plugins/use-features
*
* @example
* ```ts
* import { createFeatures } from '@vuetify/v0'
*
* const [useFeatures, provideFeaturesContext, context] = createFeatures({
*   namespace: 'v0:features',
*   features: {
*     'dark-mode': true,
*     'theme-color': { $variation: 'blue' },
*   },
* })
* ```
*/
function createFeatures(_options = {}) {
	const { features, ...options } = _options;
	const tokens = createTokens(features, { flat: true });
	const registry = createGroup({
		...options,
		reactive: true
	});
	for (const [id, { value }] of tokens.entries()) register({
		id,
		value
	});
	function variation(id, fallback = null) {
		const ticket = registry.get(id);
		if (!ticket) return fallback;
		return /* @__PURE__ */ isObject(ticket.value) ? ticket.value.$variation ?? fallback : ticket.value ?? fallback;
	}
	function register(registration = {}) {
		const item = {
			value: false,
			...registration
		};
		const ticket = registry.register(item);
		if (/* @__PURE__ */ isBoolean(ticket.value) && ticket.value === true || /* @__PURE__ */ isObject(ticket.value) && /* @__PURE__ */ isBoolean(ticket.value.$value) && ticket.value.$value === true) registry.select(ticket.id);
		return ticket;
	}
	function sync(flags) {
		for (const [id, value] of Object.entries(flags)) if (registry.get(id)) {
			const shouldSelect = /* @__PURE__ */ isBoolean(value) ? value === true : /* @__PURE__ */ isObject(value) && /* @__PURE__ */ isBoolean(value.$value) && value.$value === true;
			registry.upsert(id, { value });
			if (shouldSelect) registry.select(id);
			else registry.unselect(id);
		} else register({
			id,
			value
		});
	}
	function onboard(registrations) {
		return registry.batch(() => registrations.map((registration) => register(registration)));
	}
	return {
		...registry,
		variation,
		register,
		onboard,
		sync,
		get size() {
			return registry.size;
		}
	};
}
function createFeaturesFallback() {
	return {
		size: 0,
		variation: (_id, fallback = null) => fallback,
		sync: () => {}
	};
}
const [createFeaturesContext, createFeaturesPlugin, useFeatures] = createPluginContext("v0:features", (options) => createFeatures(options), {
	fallback: () => createFeaturesFallback(),
	setup: (context, app, { adapter }) => {
		if (!adapter) return;
		for (const a of /* @__PURE__ */ isArray(adapter) ? adapter : [adapter]) {
			context.sync(a.setup((flags) => context.sync(flags)));
			if (/* @__PURE__ */ isFunction(a.dispose)) app.onUnmount(() => a.dispose());
		}
	}
});
//#endregion
export { FeaturesAdapter as a, useFeatures as i, createFeaturesContext as n, createFeaturesPlugin as r, createFeatures as t };
