// @vuetify/v0/dist/useLogger-C3crtWEd.mjs
import { s as __LOGGER_ENABLED__, t as IN_BROWSER } from "./globals-DCznFHqN.mjs";
import { isUndefined } from "./utilities/index.mjs";
import { i as useContext, n as createContext, t as createTrinity } from "./createTrinity-C_c8eedz.mjs";
import { t as Vuetify0LoggerAdapter } from "./v0-2IcFDTW5.mjs";
import { hasInjectionContext } from "vue";
//#region src/composables/createPlugin/index.ts
/**
* @module createPlugin
*
* @see https://vuejs.org/guide/reusability/plugins.html
* @see https://0.vuetifyjs.com/composables/foundation/create-plugin
*
* @remarks
* Factory for creating Vue plugins with proper context provision.
*
* Wraps the provide function in app.runWithContext() to ensure proper execution context,
* allowing plugins to safely provide dependency injection contexts at the application level.
*
* Also exports `createPluginContext` — a higher-level factory that generates the standard
* context/plugin/consumer triple for plugin composables, eliminating boilerplate.
*/
const INSTALLED = Symbol.for("v0:installed-plugins");
function getInstalled(app) {
	const ctx = app._context;
	return ctx[INSTALLED] ??= /* @__PURE__ */ new Set();
}
/**
* Creates a new Vue plugin.
*
* @param options The plugin options.
* @returns A new Vue plugin.
*
* @see https://0.vuetifyjs.com/composables/foundation/create-plugin#create-plugin
*
* @example
* ```ts
* export const [useContext, provideContext] = createContext<MyContext>('my-plugin')
*
* const context = {}
*
* export const MyPlugin = createPlugin({
*   namespace: 'my-plugin',
*   provide: (app) => {
*     provideContext(context, app)
*   },
*   setup: (app) => {
*     // Optional setup logic
*   },
* })
*/
function createPlugin(options) {
	return { install(app) {
		app.runWithContext(() => {
			const installed = getInstalled(app);
			if (installed.has(options.namespace)) return;
			installed.add(options.namespace);
			options.provide(app);
			options.setup?.(app);
		});
	} };
}
/**
* Creates the three standard functions for a plugin composable.
*
* @param defaultNamespace The default DI namespace string (e.g. `'v0:logger'`).
* @param factory Function that creates the composable context instance from options.
* @param config Optional setup callback and fallback factory.
* @returns A readonly tuple: `[createXContext, createXPlugin, useX]`.
*
* @example
* ```ts
* // Simple — no setup or fallback
* export const [createStorageContext, createStoragePlugin, useStorage] =
*   createPluginContext('v0:storage', options => createStorage(options))
*
* // With fallback — safe outside component instances
* export const [createLoggerContext, createLoggerPlugin, useLogger] =
*   createPluginContext('v0:logger', options => createLogger(options), {
*     fallback: ns => createFallbackLogger(ns),
*     setup: (context) => {
*       if (__DEV__ && IN_BROWSER) (window as any).__v0Logger__ = context
*     },
*   })
* ```
*/
function createPluginContext(defaultNamespace, factory, config) {
	function createXContext(_options = {}) {
		const { namespace = defaultNamespace, ...options } = _options;
		const [_use, _provide] = createContext(namespace);
		const context = factory(options);
		function provide(_context = context, app) {
			return _provide(_context, app);
		}
		return createTrinity(_use, provide, context);
	}
	function createXPlugin(_options = {}) {
		const { namespace = defaultNamespace, ...options } = _options;
		const [, provide, context] = createXContext({
			...options,
			namespace
		});
		return createPlugin({
			namespace,
			provide: (app) => {
				provide(context, app);
			},
			setup: config?.setup ? (app) => config.setup(context, app, options) : void 0
		});
	}
	function useX(namespace = defaultNamespace) {
		if (config?.fallback) {
			const instance = config.fallback(namespace);
			if (!hasInjectionContext()) return instance;
			return useContext(namespace, instance);
		}
		return useContext(namespace);
	}
	return [
		createXContext,
		createXPlugin,
		useX
	];
}
//#endregion
//#region src/composables/useLogger/index.ts
/**
* @module useLogger
*
* @see https://0.vuetifyjs.com/composables/plugins/use-logger
*
* @remarks
* Logging composable with adapter pattern supporting console, consola, and pino.
*
* Key features:
* - Multiple log levels (trace, debug, info, warn, error, fatal)
* - Adapter pattern for console/consola/pino integration
* - Enable/disable logging
* - Fallback logger for undefined loggers
* - Context logging support
*
* Uses adapter pattern to abstract logging implementation.
*/
/**
* Creates a new logger instance.
*
* @param options The options for the logger instance.
* @returns A new logger instance.
*
* @see https://0.vuetifyjs.com/composables/plugins/use-logger
*
* @example
* ```ts
* import { createLogger } from '@vuetify/v0'
*
* const logger = createLogger({
*   level: 'debug',
*   prefix: '[MyApp]',
* })
*
* logger.info('This is an info message')
* logger.debug('This is a debug message')
* logger.error('This is an error message')
* logger.level('debug')
* logger.debug('This debug message will now be logged')
* ```
*/
function createLogger(options = {}) {
	const { adapter = new Vuetify0LoggerAdapter({ prefix: options.prefix }), level: initialLevel = "info", enabled: initialEnabled = __LOGGER_ENABLED__ } = options;
	let currentLevel = initialLevel;
	let isEnabled = initialEnabled;
	function value(level) {
		return {
			trace: 0,
			debug: 1,
			info: 2,
			warn: 3,
			error: 4,
			fatal: 5,
			silent: 6
		}[level] ?? 2;
	}
	function can(level) {
		if (!isEnabled) return false;
		return value(level) >= value(currentLevel);
	}
	function format(message) {
		return message;
	}
	function debug(message, ...args) {
		if (can("debug")) adapter.debug(format(message), ...args);
	}
	function info(message, ...args) {
		if (can("info")) adapter.info(format(message), ...args);
	}
	function warn(message, ...args) {
		if (can("warn")) adapter.warn(format(message), ...args);
	}
	function error(message, ...args) {
		if (can("error")) adapter.error(format(message), ...args);
	}
	function trace(message, ...args) {
		if (can("trace")) adapter.trace?.(format(message), ...args);
	}
	function fatal(message, ...args) {
		if (can("fatal")) adapter.fatal?.(format(message), ...args);
	}
	function level(newLevel) {
		currentLevel = newLevel;
	}
	function current() {
		return currentLevel;
	}
	function enabled() {
		return isEnabled;
	}
	function enable() {
		isEnabled = true;
	}
	function disable() {
		isEnabled = false;
	}
	return {
		debug,
		info,
		warn,
		error,
		trace,
		fatal,
		level,
		current,
		enabled,
		enable,
		disable
	};
}
function createFallbackLogger(namespace = "v0:logger") {
	function format(message, type) {
		return `[${namespace} ${type}] ${message}`;
	}
	return {
		debug: (message, ...args) => console.debug(format(message, "debug"), ...args),
		info: (message, ...args) => console.info(format(message, "info"), ...args),
		warn: (message, ...args) => console.warn(format(message, "warn"), ...args),
		error: (message, ...args) => console.error(format(message, "error"), ...args),
		trace: (message, ...args) => console.trace(format(message, "trace"), ...args),
		fatal: (message, ...args) => console.error(format(message, "fatal"), ...args),
		level: () => {},
		current: () => "info",
		enabled: () => true,
		enable: () => {},
		disable: () => {}
	};
}
/**
* Creates a new logger context.
*
* @param options The options for the logger context.
* @template E The type of the logger context.
* @returns A new logger context.
*
* @see https://0.vuetifyjs.com/composables/plugins/use-logger
*
* @example
* ```ts
* import { createLoggerContext } from '@vuetify/v0'
*
* export const [useAppLogger, provideAppLogger, appLogger] = createLoggerContext({
*   namespace: 'app:logger',
*   level: 'debug',
* })
* ```
*/
const [createLoggerContext, createLoggerPlugin, useLogger] = createPluginContext("v0:logger", (options) => createLogger(options), {
	fallback: (ns) => createFallbackLogger(ns),
	setup: (context, _app, _options) => {
		if (!/* @__PURE__ */ isUndefined(process.env.NODE_ENV !== "production") && process.env.NODE_ENV !== "production" && IN_BROWSER) window.__v0Logger__ = context;
	}
});
//#endregion
export { createPlugin as a, useLogger as i, createLoggerContext as n, createPluginContext as o, createLoggerPlugin as r, createLogger as t };
