// @vuetify/v0/dist/useLocale-D3btFSZv.mjs
import { o as createPluginContext } from "./useLogger-C3crtWEd.mjs";
import { r as flatten, t as createTokens } from "./createTokens-CpapxkYX.mjs";
import { t as createSingle } from "./createSingle-BBwE5j_k.mjs";
import { t as Vuetify0LocaleAdapter } from "./v0-CB-GCB-s.mjs";
//#region src/composables/useLocale/index.ts
/**
* @module useLocale
*
* @remarks
* Internationalization (i18n) composable with adapter pattern for message translation.
*
* Key features:
* - Locale selection with createSingle
* - Token-based message storage with useTokens
* - Numbered and named placeholder support ({0}, {name})
* - Number formatting with Intl.NumberFormat
* - Adapter pattern for integration with i18n providers
*
* Integrates with createSingle for locale selection and useTokens for message resolution.
*/
/**
* Creates a new locale instance.
*
* @param options The options for the locale instance.
* @template Z The type of the locale ticket.
* @template E The type of the locale context.
* @returns A new locale instance.
*
* @see https://0.vuetifyjs.com/composables/plugins/use-locale
*/
function createLocale(_options = {}) {
	const { adapter: externalAdapter, messages = {}, fallback: fallbackLocale, ...options } = _options;
	const tokens = createTokens(messages);
	const registry = createSingle(options);
	for (const id in messages) {
		registry.register({ id });
		if (id === options.default && !registry.selectedId.value) registry.select(id);
	}
	const adapter = externalAdapter ?? new Vuetify0LocaleAdapter({
		tokens,
		selectedId: registry.selectedId,
		fallbackLocale,
		has: (id) => registry.has(id)
	});
	function t(key, ...params) {
		return adapter.t(key, ...params);
	}
	function n(value) {
		return adapter.n(value);
	}
	function register(registration = {}) {
		const { messages: msgs, ...rest } = registration;
		if (msgs && rest.id && !registry.has(rest.id)) tokens.onboard(flatten({ [rest.id]: msgs }));
		return registry.register(rest);
	}
	return {
		...registry,
		t,
		n,
		register,
		get size() {
			return registry.size;
		}
	};
}
function createLocaleFallback() {
	return {
		size: 0,
		t: (key) => key,
		n: String
	};
}
const [createLocaleContext, createLocalePlugin, useLocale] = createPluginContext("v0:locale", (options) => createLocale(options), { fallback: () => createLocaleFallback() });
//#endregion
export { useLocale as a, createLocalePlugin as i, createLocaleContext as n, createLocaleFallback as r, createLocale as t };
