// @vuetify/v0/dist/usePermissions-DmaYLhw5.mjs
import { o as createPluginContext } from "./useLogger-C3crtWEd.mjs";
import { t as createTokens } from "./createTokens-CpapxkYX.mjs";
import { t as toArray } from "./toArray-VdROPpy-.mjs";
import { t as Vuetify0PermissionAdapter } from "./v0-B1QQGE4O.mjs";
//#region src/composables/usePermissions/index.ts
/**
* @module usePermissions
*
* @see https://0.vuetifyjs.com/composables/plugins/use-permissions
*
* @remarks
* Permission management composable with support for RBAC and ABAC patterns.
*
* Key features:
* - Role-Based Access Control (RBAC) support
* - Attribute-Based Access Control (ABAC) with context
* - Functional permission conditions
* - Token-based permission storage
* - Adapter pattern for custom permission systems
*
* Built on useTokens for flexible permission configuration.
*/
/**
* Creates a new permissions instance.
*
* @param options The options for the permissions instance.
* @template Z The type of the permission ticket.
* @template E The type of the permission context.
* @returns A new permissions instance.
*
* @see https://0.vuetifyjs.com/composables/plugins/use-permissions
*
* @example
* ```ts
* import { createPermissions } from '@vuetify/v0'
*
* const [usePermissions, providePermissions] = createPermissions({
*   namespace: 'v0:permissions',
*   permissions: {
*     admin: [['read', 'users']],
*     editor: [['edit', 'posts']],
*   },
* })
* ```
*/
function createPermissions(_options = {}) {
	const { adapter = new Vuetify0PermissionAdapter(), permissions = {}, ...options } = _options;
	const record = {};
	for (const role in permissions) {
		if (!record[role]) record[role] = {};
		for (const [actions, subjects, condition = true] of permissions[role]) for (const action of toArray(actions)) for (const subject of toArray(subjects)) {
			if (!record[role][action]) record[role][action] = {};
			record[role][action][subject] = condition;
		}
	}
	const tokens = createTokens(record, options);
	function can(id, action, subject, context = {}) {
		return adapter.can(id, action, subject, context, tokens);
	}
	return {
		...tokens,
		can
	};
}
/**
* Creates a new permissions context.
*
* @param options The options for the permissions context.
* @template Z The type of the permission ticket.
* @template E The type of the permission context.
* @returns A new permissions context.
*
* @see https://0.vuetifyjs.com/composables/plugins/use-permissions
*
* @example
* ```ts
* import { createPermissionsContext } from '@vuetify/v0'
*
* export const [usePermissions, providePermissions, context] = createPermissionsContext({
*   namespace: 'app:permissions',
*   permissions: {
*     admin: [['read', 'users'], ['edit', 'users']],
*     editor: [['edit', 'posts']],
*   },
* })
* ```
*/
function createPermissionsFallback() {
	return {
		size: 0,
		can: () => false
	};
}
const [createPermissionsContext, createPermissionsPlugin, usePermissions] = createPluginContext("v0:permissions", (options) => createPermissions(options), { fallback: () => createPermissionsFallback() });
/**
* Creates a new permissions plugin.
*
* @param options The options for the permissions plugin.
* @template Z The type of the permission ticket.
* @template E The type of the permission context.
* @returns A new permissions plugin.
*
* @see https://0.vuetifyjs.com/composables/plugins/use-permissions
*
* @example
* ```ts
* import { createApp } from 'vue'
* import { createPermissionsPlugin } from '@vuetify/v0'
* import App from './App.vue'
*
* const app = createApp(App)
*
* app.use(
*   createPermissionsPlugin({
*     permissions: {
*       admin: [['read', 'users']],
*       editor: [['edit', 'posts']],
*     },
*   })
* )
*
* app.mount('#app')
* ```
*/
//#endregion
export { usePermissions as i, createPermissionsContext as n, createPermissionsPlugin as r, createPermissions as t };
