// @vuetify/v0/dist/useTheme-MJA1p7YK.mjs
import { foreground } from "./utilities/index.mjs";
import { o as createPluginContext } from "./useLogger-C3crtWEd.mjs";
import { r as flatten, t as createTokens } from "./createTokens-CpapxkYX.mjs";
import { t as createSingle } from "./createSingle-BBwE5j_k.mjs";
import { t as Vuetify0ThemeAdapter } from "./v0-pzmWhctY.mjs";
import { computed, shallowRef, toRef } from "vue";
//#region src/composables/useTheme/index.ts
/**
* @module useTheme
*
* @see https://0.vuetifyjs.com/composables/plugins/use-theme
*
* @remarks
* Theme management composable with token resolution and CSS variable injection.
*
* Key features:
* - Single-selection theme switching (extends createSingle)
* - Token alias resolution via useTokens
* - Lazy theme loading (compute colors only when selected)
* - CSS variable generation via adapter pattern
* - SSR support with head integration
* - Theme cycling
*
* Integrates with createSingle for selection and useTokens for color resolution.
*/
/**
* Creates a new theme instance.
*
* @param options The options for the theme instance.
* @template Z The type of the theme ticket.
* @template E The type of the theme context.
* @returns A new theme instance.
*
* @see https://0.vuetifyjs.com/composables/plugins/use-theme
*
* @example
* ```ts
* import { createTheme } from '@vuetify/v0'
*
* export const [useTheme, provideTheme] = createTheme({
*   namespace: 'v0:theme',
*   default: 'light',
*   themes: {
*     light: {
*       dark: false,
*       colors: {
*         primary: '#3b82f6',
*       },
*     },
*     dark: {
*       dark: true,
*       colors: {
*         primary: '#675496',
*       },
*     },
*   },
* })
* ```
*/
function createTheme(_options = {}) {
	const { themes = {}, palette = {}, foreground: genForeground, ...options } = _options;
	const tokens = createTokens({
		palette,
		...themes
	}, { flat: true });
	const registry = createSingle(options);
	for (const id in themes) {
		const { colors: value, ...theme } = themes[id];
		register({
			id,
			value,
			...theme
		});
		if (id === options.default && !registry.selectedId.value) registry.select(id);
	}
	const names = toRef(() => registry.keys());
	const colors = computed(() => {
		const resolved = {};
		for (const theme of registry.values()) {
			if (theme.lazy && theme.id !== registry.selectedId.value) continue;
			const themeColors = resolve(theme.value);
			if (genForeground) for (const [key, value] of Object.entries(themeColors)) {
				const onKey = `on-${key}`;
				if (!key.startsWith("on-") && !(onKey in themeColors)) themeColors[onKey] = foreground(value);
			}
			resolved[theme.id] = themeColors;
		}
		return resolved;
	});
	const isDark = toRef(() => registry.selectedItem.value?.dark ?? false);
	function cycle(themes = names.value) {
		const current = themes.indexOf(registry.selectedId.value ?? "");
		const next = current === -1 ? 0 : (current + 1) % themes.length;
		registry.select(themes[next]);
	}
	function resolve(colors) {
		const resolved = {};
		for (const [key, value] of Object.entries(colors)) resolved[key] = tokens.isAlias(value) ? tokens.resolve(value) : value;
		return resolved;
	}
	function register(registration = {}) {
		const { colors, ...rest } = registration;
		if (colors && rest.id && !registry.has(rest.id)) tokens.onboard(flatten({ [rest.id]: { colors } }, "", true));
		const item = {
			lazy: false,
			dark: false,
			...rest,
			...colors ? { value: colors } : {}
		};
		return registry.register(item);
	}
	function onboard(registrations) {
		return registry.batch(() => registrations.map((registration) => register(registration)));
	}
	return {
		...registry,
		colors,
		isDark,
		register,
		onboard,
		cycle,
		get size() {
			return registry.size;
		}
	};
}
function createThemeFallback() {
	return {
		size: 0,
		colors: computed(() => ({})),
		isDark: shallowRef(false),
		cycle: () => {}
	};
}
const [createThemeContext, createThemePlugin, useTheme] = createPluginContext("v0:theme", (options) => createTheme(options), {
	fallback: () => createThemeFallback(),
	setup: (context, app, { adapter = new Vuetify0ThemeAdapter(), target, rgb }) => {
		if (rgb) adapter.rgb = true;
		adapter.setup(app, context, target);
	}
});
//#endregion
export { useTheme as i, createThemeContext as n, createThemePlugin as r, createTheme as t };
