// @vuetify/v0/dist/v0-pzmWhctY.mjs
import { t as IN_BROWSER } from "./globals-DCznFHqN.mjs";
import { hexToRgb, isNull, isString, isUndefined } from "./utilities/index.mjs";
import { onScopeDispose, watch } from "vue";
//#region src/composables/useTheme/adapters/adapter.ts
var ThemeAdapter = class {
	stylesheetId = "v0-theme-stylesheet";
	prefix;
	rgb = false;
	constructor(prefix) {
		this.prefix = prefix;
	}
	generate(colors, isDark) {
		let css = "";
		for (const theme in colors) {
			const themeColors = colors[theme];
			if (!themeColors) continue;
			const vars = Object.entries(themeColors).map(([key, val]) => `  --${this.prefix}-${key}: ${this.rgb ? this.decompose(val) : val};`).join("\n");
			css += `[data-theme="${theme}"] {\n${vars}\n}\n`;
		}
		if (!/* @__PURE__ */ isUndefined(isDark)) css += `:root {\n  color-scheme: ${isDark ? "dark" : "light"};\n}\n`;
		return css;
	}
	decompose(hex) {
		const { r, g, b, a } = hexToRgb(hex);
		return /* @__PURE__ */ isUndefined(a) ? `${r}, ${g}, ${b}` : `${r}, ${g}, ${b}, ${a}`;
	}
};
//#endregion
//#region src/composables/useTheme/adapters/v0.ts
/**
* Theme adapter implementation for Vuetify v0 design system.
* This adapter generates CSS custom properties and injects them into the DOM
* as a stylesheet, allowing themes to be applied globally.
*/
var Vuetify0ThemeAdapter = class extends ThemeAdapter {
	cspNonce;
	sheet;
	constructor(options = {}) {
		super(options.prefix ?? "v0");
		this.cspNonce = options.cspNonce;
		this.stylesheetId = options.stylesheetId ?? this.stylesheetId;
	}
	setup(app, context, target) {
		if (IN_BROWSER) {
			this.update(context.colors.value, context.isDark.value);
			onScopeDispose(watch([context.colors, context.isDark], ([colors, isDark]) => {
				this.update(colors, isDark);
			}), true);
			if (/* @__PURE__ */ isNull(target)) return;
			const targetEl = target instanceof HTMLElement ? target : /* @__PURE__ */ isString(target) ? document.querySelector(target) : app._container || document.querySelector("#app") || document.body;
			if (!targetEl) return;
			if (context.selectedId.value) targetEl.dataset.theme = String(context.selectedId.value);
			onScopeDispose(watch(context.selectedId, (id) => {
				if (!id) return;
				targetEl.dataset.theme = String(id);
			}), true);
		} else {
			const head = app._context?.provides?.usehead ?? app._context?.provides?.head;
			if (head?.push) {
				const id = context.selectedId.value;
				head.push({
					htmlAttrs: { "data-theme": id ? String(id) : "" },
					style: [{
						innerHTML: this.generate(context.colors.value, context.isDark.value),
						id: this.stylesheetId
					}]
				});
			}
		}
	}
	update(colors, isDark) {
		if (!IN_BROWSER) return;
		this.upsert(this.generate(colors, isDark));
	}
	upsert(styles) {
		if (!IN_BROWSER) return;
		if (!this.sheet) {
			this.sheet = new CSSStyleSheet();
			document.adoptedStyleSheets = [...document.adoptedStyleSheets, this.sheet];
		}
		this.sheet.replaceSync(styles);
	}
};
//#endregion
export { ThemeAdapter as n, Vuetify0ThemeAdapter as t };
