// @vuetify/v0/dist/notifications/index.mjs
import { t as IN_BROWSER } from "../globals-DCznFHqN.mjs";
//#region src/composables/useNotifications/adapters/knock.ts
/**
* @module KnockNotificationsAdapter
*
* @remarks
* Knock adapter for useNotifications.
* Bridges a Knock feed instance with the notification system.
*
* Inbound: Maps real-time feed events to ctx.send()
* Outbound: Listens to notification:read and notification:archived
*           events and calls feed.markAsRead/markAsArchived
*
* @example
* ```ts
* import Knock from '@knocklabs/client'
* import { createKnockAdapter } from '#v0/composables/useNotifications/adapters/knock'
*
* const knock = new Knock(KNOCK_PUBLIC_API_KEY)
* knock.authenticate(userId)
* const feed = knock.feeds.initialize(KNOCK_FEED_CHANNEL_ID)
*
* app.use(createNotificationsPlugin({
*   adapter: createKnockAdapter(feed),
* }))
* ```
*/
function noop$1() {}
function mapItem$1(item) {
	const subject = item.blocks?.find((b) => b.name === "subject")?.rendered;
	const body = item.blocks?.find((b) => b.name === "body")?.rendered;
	return {
		id: item.id,
		subject,
		body,
		data: item.data
	};
}
function createKnockAdapter(feed) {
	const items = /* @__PURE__ */ new Map();
	let ctx;
	let onReceived;
	let onPage;
	let onRead;
	let onArchived;
	return {
		setup(_ctx) {
			ctx = _ctx;
			onReceived = (data) => {
				const payload = data;
				if (!payload?.items) return;
				for (const item of payload.items) {
					if (items.has(item.id)) continue;
					items.set(item.id, item);
					ctx.send(mapItem$1(item));
				}
			};
			onPage = (data) => {
				const payload = data;
				if (!payload?.items) return;
				for (const item of payload.items) {
					if (items.has(item.id)) continue;
					items.set(item.id, item);
					ctx.register(mapItem$1(item));
				}
			};
			feed.on("items.received.realtime", onReceived);
			feed.on("items.received.page", onPage);
			if (IN_BROWSER) {
				feed.listenForUpdates();
				feed.fetch().catch(noop$1);
			}
			onRead = (data) => {
				const item = items.get(String(data));
				if (item) feed.markAsRead(item).catch(noop$1);
			};
			onArchived = (data) => {
				const item = items.get(String(data));
				if (item) feed.markAsArchived(item).catch(noop$1);
			};
			ctx.on("notification:read", onRead);
			ctx.on("notification:archived", onArchived);
		},
		dispose() {
			if (!ctx) return;
			feed.off("items.received.realtime", onReceived);
			feed.off("items.received.page", onPage);
			ctx.off("notification:read", onRead);
			ctx.off("notification:archived", onArchived);
			items.clear();
			feed.teardown();
		}
	};
}
//#endregion
//#region src/composables/useNotifications/adapters/novu.ts
/**
* @module NovuNotificationsAdapter
*
* @remarks
* Novu adapter for useNotifications.
* Bridges a Novu client instance with the notification system.
*
* Inbound: Seeds registry from novu.notifications.list() on setup,
*          then listens to notifications.notification_received for real-time updates.
* Outbound: Listens to notification lifecycle events and calls the Novu API
*           (read, unread, seen, archive, unarchive).
*
* @example
* ```ts
* import { Novu } from '@novu/js'
* import { createNovuAdapter } from '@vuetify/v0/notifications'
*
* const novu = new Novu({
*   subscriberId: userId,
*   applicationIdentifier: import.meta.env.VITE_NOVU_APP_ID,
* })
*
* app.use(createNotificationsPlugin({
*   adapter: createNovuAdapter(novu),
* }))
* ```
*/
const DEFAULT_SEVERITY_MAP = {
	critical: "error",
	high: "error",
	medium: "warning",
	low: "info"
};
function defaultSeverity(s) {
	return DEFAULT_SEVERITY_MAP[s.toLowerCase()];
}
function noop() {}
function mapItem(item, resolveSeverity) {
	return {
		id: item.id,
		subject: item.subject,
		body: item.body,
		data: item.data,
		severity: item.severity ? resolveSeverity(item.severity) : void 0
	};
}
function createNovuAdapter(novu, options = {}) {
	const resolveSeverity = options.severity ?? defaultSeverity;
	const ids = /* @__PURE__ */ new Set();
	let ctx;
	let disposed = false;
	const unsubscribers = [];
	let onRead;
	let onUnread;
	let onSeen;
	let onArchived;
	let onUnarchived;
	return {
		setup(_ctx) {
			ctx = _ctx;
			disposed = false;
			const unsub = novu.on("notifications.notification_received", (data) => {
				const item = data;
				if (disposed || !item?.id || ids.has(item.id)) return;
				ids.add(item.id);
				ctx.send(mapItem(item, resolveSeverity));
			});
			unsubscribers.push(unsub);
			if (IN_BROWSER) novu.notifications.list({ limit: 30 }).then(({ data }) => {
				if (disposed) return;
				for (const item of data?.notifications ?? []) {
					if (!item.id || ids.has(item.id)) continue;
					ids.add(item.id);
					ctx.register(mapItem(item, resolveSeverity));
				}
			}).catch(noop);
			onRead = (data) => {
				const id = String(data);
				if (ids.has(id)) novu.notifications.read({ notificationId: id }).catch(noop);
			};
			onUnread = (data) => {
				const id = String(data);
				if (ids.has(id)) novu.notifications.unread({ notificationId: id }).catch(noop);
			};
			onSeen = (data) => {
				const id = String(data);
				if (ids.has(id)) novu.notifications.seenAll({ notificationIds: [id] }).catch(noop);
			};
			onArchived = (data) => {
				const id = String(data);
				if (ids.has(id)) novu.notifications.archive({ notificationId: id }).catch(noop);
			};
			onUnarchived = (data) => {
				const id = String(data);
				if (ids.has(id)) novu.notifications.unarchive({ notificationId: id }).catch(noop);
			};
			ctx.on("notification:read", onRead);
			ctx.on("notification:unread", onUnread);
			ctx.on("notification:seen", onSeen);
			ctx.on("notification:archived", onArchived);
			ctx.on("notification:unarchived", onUnarchived);
		},
		dispose() {
			disposed = true;
			for (const unsub of unsubscribers) unsub();
			unsubscribers.length = 0;
			if (ctx) {
				if (onRead) ctx.off("notification:read", onRead);
				if (onUnread) ctx.off("notification:unread", onUnread);
				if (onSeen) ctx.off("notification:seen", onSeen);
				if (onArchived) ctx.off("notification:archived", onArchived);
				if (onUnarchived) ctx.off("notification:unarchived", onUnarchived);
			}
			ids.clear();
		}
	};
}
//#endregion
export { createKnockAdapter, createNovuAdapter };
