// @vuetify/v0/dist/utilities/index.mjs
import { t as IN_BROWSER } from "../globals-DCznFHqN.mjs";
import * as Vue from "vue";
import { useId as useId$1 } from "vue";
//#region src/utilities/color.ts
function hexToRgb(hex) {
	const clean = hex.startsWith("#") ? hex.slice(1) : hex;
	if (clean.length === 3 || clean.length === 4) {
		const r = Number.parseInt(clean[0] + clean[0], 16);
		const g = Number.parseInt(clean[1] + clean[1], 16);
		const b = Number.parseInt(clean[2] + clean[2], 16);
		const a = clean.length === 4 ? Number.parseInt(clean[3] + clean[3], 16) : void 0;
		return a === void 0 ? {
			r,
			g,
			b
		} : {
			r,
			g,
			b,
			a
		};
	}
	const r = Number.parseInt(clean.slice(0, 2), 16);
	const g = Number.parseInt(clean.slice(2, 4), 16);
	const b = Number.parseInt(clean.slice(4, 6), 16);
	const a = clean.length === 8 ? Number.parseInt(clean.slice(6, 8), 16) : void 0;
	return a === void 0 ? {
		r,
		g,
		b
	} : {
		r,
		g,
		b,
		a
	};
}
function rgbToHex({ r, g, b, a }) {
	const hex = "#" + [
		r,
		g,
		b
	].map((c) => c.toString(16).padStart(2, "0")).join("");
	if (a !== void 0) return hex + a.toString(16).padStart(2, "0");
	return hex;
}
//#endregion
//#region src/utilities/apca.ts
const Rco = .2126729;
const Gco = .7151522;
const Bco = .072175;
const normBG = .56;
const normTXT = .57;
const revBG = .65;
const revTXT = .62;
const blkThrs = .022;
const blkClmp = 1.414;
const scaleBoW = 1.14;
const scaleWoB = 1.14;
const loBoWoffset = .027;
const loWoBoffset = .027;
const deltaYmin = 5e-4;
function linearize(channel) {
	const val = channel / 255;
	return val <= .04045 ? val / 12.92 : Math.pow((val + .055) / 1.055, 2.4);
}
function luminance(rgb) {
	return Rco * linearize(rgb.r) + Gco * linearize(rgb.g) + Bco * linearize(rgb.b);
}
function clamp$1(y) {
	return y > blkThrs ? y : y + Math.pow(blkThrs - y, blkClmp);
}
function apca(text, background) {
	const txtY = clamp$1(luminance(text));
	const bgY = clamp$1(luminance(background));
	if (Math.abs(bgY - txtY) < deltaYmin) return 0;
	let contrast;
	if (bgY > txtY) {
		contrast = (Math.pow(bgY, normBG) - Math.pow(txtY, normTXT)) * scaleBoW;
		return contrast < loBoWoffset ? 0 : (contrast - loBoWoffset) * 100;
	} else {
		contrast = (Math.pow(bgY, revBG) - Math.pow(txtY, revTXT)) * scaleWoB;
		return contrast > -loWoBoffset ? 0 : (contrast + loWoBoffset) * 100;
	}
}
const BLACK = {
	r: 0,
	g: 0,
	b: 0
};
const WHITE = {
	r: 255,
	g: 255,
	b: 255
};
function foreground(hex) {
	const rgb = hexToRgb(hex);
	const onWhite = Math.abs(apca(WHITE, rgb));
	return Math.abs(apca(BLACK, rgb)) > onWhite ? "#000000" : "#ffffff";
}
//#endregion
//#region src/utilities/instance.ts
/**
* Vue 3.6+ exposes `currentInstance` for Vapor mode compatibility.
* `getCurrentInstance()` returns null in Vapor components by design,
* so we access `currentInstance` directly when available.
*
* Uses dynamic key to avoid bundler static analysis warnings.
*
* @see https://github.com/orgs/vuejs/discussions/13629
* @internal
*/
const INSTANCE_KEY = "currentInstance";
function getInstanceCompat() {
	const vueExports = Vue;
	if (INSTANCE_KEY in vueExports) return vueExports[INSTANCE_KEY];
	return Vue.getCurrentInstance();
}
function instanceExists() {
	return !/* @__PURE__ */ isNull(getInstanceCompat());
}
function instanceName() {
	return getInstanceCompat()?.type?.name;
}
//#endregion
//#region src/utilities/helpers.ts
/**
* @module utilities/helpers
*
* @remarks
* Collection of utility functions for type checking, object manipulation,
* and common transformations. All exports are tree-shakeable.
*/
/**
* Checks if a value is a function
*
* @param item The value to check
* @returns True if the value is a function
*
* @example
* ```ts
* isFunction(() => {})  // true
* isFunction('string')  // false
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function isFunction(item) {
	return typeof item === "function";
}
/**
* Checks if a value is a string
*
* @param item The value to check
* @returns True if the value is a string
*
* @example
* ```ts
* isString('hello')  // true
* isString(123)      // false
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function isString(item) {
	return typeof item === "string";
}
/**
* Checks if a value is a number
*
* @param item The value to check
* @returns True if the value is a number (including NaN)
*
* @example
* ```ts
* isNumber(123)       // true
* isNumber(NaN)       // true
* isNumber('123')     // false
* ```
*
* @see {@link isNaN} to check for NaN specifically
*/
/* @__NO_SIDE_EFFECTS__ */
function isNumber(item) {
	return typeof item === "number";
}
/**
* Checks if a value is a boolean
*
* @param item The value to check
* @returns True if the value is a boolean
*
* @example
* ```ts
* isBoolean(true)   // true
* isBoolean(false)  // true
* isBoolean(0)      // false
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function isBoolean(item) {
	return typeof item === "boolean";
}
/**
* Checks if a value is a plain object (excludes null and arrays)
*
* @param item The value to check
* @returns True if the value is a plain object
*
* @remarks
* Returns false for null and arrays, even though `typeof null === 'object'`
* and `typeof [] === 'object'` in JavaScript.
*
* @example
* ```ts
* isObject({})        // true
* isObject({ a: 1 })  // true
* isObject(null)      // false
* isObject([])        // false
* ```
*
* @see {@link isArray} to check for arrays
* @see {@link isNull} to check for null
*/
/* @__NO_SIDE_EFFECTS__ */
function isObject(item) {
	return typeof item === "object" && item !== null && !Array.isArray(item);
}
/**
* Checks if a value is an array
*
* @param item The value to check
* @returns True if the value is an array
*
* @example
* ```ts
* isArray([])        // true
* isArray([1, 2, 3]) // true
* isArray('string')  // false
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function isArray(item) {
	return Array.isArray(item);
}
/**
* Checks if a value is a DOM Element
*
* @param item The value to check
* @returns True if the value is a DOM Element
*
* @example
* ```ts
* isElement(document.body) // true
* isElement('string')      // false
* isElement(null)          // false
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function isElement(item) {
	return item instanceof Element;
}
/**
* Checks if a value is null
*
* @param item The value to check
* @returns True if the value is null
*
* @example
* ```ts
* isNull(null)      // true
* isNull(undefined) // false
* ```
*
* @see {@link isUndefined} to check for undefined
* @see {@link isNullOrUndefined} to check for either
*/
/* @__NO_SIDE_EFFECTS__ */
function isNull(item) {
	return item === null;
}
/**
* Checks if a value is null or undefined
*
* @param item The value to check
* @returns True if the value is null or undefined
*
* @remarks
* Uses loose equality (`== null`) which matches both null and undefined.
*
* @example
* ```ts
* isNullOrUndefined(null)      // true
* isNullOrUndefined(undefined) // true
* isNullOrUndefined(0)         // false
* isNullOrUndefined('')        // false
* ```
*
* @see {@link isNull} to check for null only
* @see {@link isUndefined} to check for undefined only
*/
/* @__NO_SIDE_EFFECTS__ */
function isNullOrUndefined(item) {
	return item == null;
}
/**
* Checks if a value is undefined
*
* @param item The value to check
* @returns True if the value is undefined
*
* @example
* ```ts
* isUndefined(undefined) // true
* isUndefined(null)      // false
* ```
*
* @see {@link isNull} to check for null
* @see {@link isNullOrUndefined} to check for either
*/
/* @__NO_SIDE_EFFECTS__ */
function isUndefined(item) {
	return item === void 0;
}
/**
* Checks if a value is a primitive (string, number, or boolean)
*
* @param item The value to check
* @returns True if the value is a string, number, or boolean
*
* @example
* ```ts
* isPrimitive('hello') // true
* isPrimitive(123)     // true
* isPrimitive(true)    // true
* isPrimitive({})      // false
* isPrimitive(null)    // false
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function isPrimitive(item) {
	return typeof item === "string" || typeof item === "number" || typeof item === "boolean";
}
/**
* Checks if a value is a symbol
*
* @param item The value to check
* @returns True if the value is a symbol
*
* @example
* ```ts
* isSymbol(Symbol('test')) // true
* isSymbol('symbol')       // false
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function isSymbol(item) {
	return typeof item === "symbol";
}
/**
* Checks if a value is NaN (Not a Number)
*
* @param item The value to check
* @returns True if the value is NaN
*
* @remarks
* Uses `Number.isNaN()` which only returns true for the actual NaN value,
* unlike the global `isNaN()` which coerces the argument to a number first.
*
* @example
* ```ts
* isNaN(NaN)         // true
* isNaN(123)         // false
* isNaN('hello')     // false (unlike global isNaN)
* isNaN(undefined)   // false (unlike global isNaN)
* ```
*
* @see {@link isNumber} to check if a value is a number type
*/
/* @__NO_SIDE_EFFECTS__ */
function isNaN(item) {
	return /* @__PURE__ */ isNumber(item) && Number.isNaN(item);
}
/**
* Deeply merges source objects into a target object, returning a new object
*
* @param target The base object to merge into
* @param sources One or more source objects to merge from
* @returns A new merged object (inputs are not mutated)
*
* @remarks
* - Creates and returns a new object — inputs are never mutated
* - Nested objects are recursively merged
* - Arrays are replaced, not merged
* - Primitives from sources overwrite target values
*
* @example
* ```ts
* const target = { a: 1, b: { c: 2 } }
* const result = mergeDeep(target, { b: { d: 3 } })
* // result is { a: 1, b: { c: 2, d: 3 } }
* // target is unchanged
*
* // Multiple sources
* mergeDeep({}, { a: 1 }, { b: 2 }) // { a: 1, b: 2 }
*
* // Arrays are replaced
* mergeDeep({ arr: [1, 2] }, { arr: [3] }) // { arr: [3] }
* ```
*/
const UNSAFE_KEYS = new Set([
	"__proto__",
	"constructor",
	"prototype"
]);
function isPlainObject(value) {
	if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
	const proto = Object.getPrototypeOf(value);
	return proto === Object.prototype || proto === null;
}
/* @__NO_SIDE_EFFECTS__ */
function mergeDeep(target, ...sources) {
	const out = {};
	for (const key in target) if (Object.prototype.hasOwnProperty.call(target, key)) out[key] = target[key];
	for (const source of sources) {
		if (!/* @__PURE__ */ isObject(source)) continue;
		for (const key in source) {
			if (UNSAFE_KEYS.has(key)) continue;
			if (!Object.prototype.hasOwnProperty.call(source, key)) continue;
			const sourceValue = source[key];
			if (/* @__PURE__ */ isUndefined(sourceValue)) continue;
			const targetValue = out[key];
			out[key] = isPlainObject(sourceValue) && isPlainObject(targetValue) ? /* @__PURE__ */ mergeDeep(targetValue, sourceValue) : sourceValue;
		}
	}
	return out;
}
let idCounter = 0;
/**
* Generates a unique ID, using Vue's useId when in component context
*
* @returns A unique string ID
*
* @remarks
* - In component setup/lifecycle: Uses Vue's `useId()` for SSR-safe hydration
* - Outside components: Falls back to sequential counter (`v0-0`, `v0-1`, ...)
* - Vapor mode compatible
*
* @example
* ```ts
* // In component setup - SSR safe
* const id = useId() // 'v:0', 'v:1', etc. (Vue's format)
*
* // Outside component - counter fallback
* const id = useId() // 'v0-0', 'v0-1', etc.
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function useId() {
	if (instanceExists()) return useId$1();
	if (!/* @__PURE__ */ isUndefined(process.env.NODE_ENV !== "production") && process.env.NODE_ENV !== "production" && !IN_BROWSER) console.warn("[v0 warn] useId() called outside component context during SSR. Provide explicit ID to avoid hydration mismatch.");
	return `v0-${idCounter++}`;
}
/**
* Clamps a value between a minimum and maximum
*
* @param value The value to clamp
* @param min The minimum value (default: 0)
* @param max The maximum value (default: 1)
* @returns The clamped value
*
* @example
* ```ts
* clamp(5, 0, 10)  // 5
* clamp(-5, 0, 10) // 0
* clamp(15, 0, 10) // 10
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function clamp(value, min = 0, max = 1) {
	return Math.max(min, Math.min(max, value));
}
/**
* Creates an array of sequential numbers
*
* @param length The length of the array to create
* @param start The starting index (default: 0)
* @returns An array of sequential numbers
*
* @example
* ```ts
* range(3)     // [0, 1, 2]
* range(3, 1)  // [1, 2, 3]
* range(5, 10) // [10, 11, 12, 13, 14]
* range(0)     // []
* ```
*/
/* @__NO_SIDE_EFFECTS__ */
function range(length, start = 0) {
	return Array.from({ length }, (_, index) => start + index);
}
//#endregion
export { apca, clamp, foreground, hexToRgb, instanceExists, instanceName, isArray, isBoolean, isElement, isFunction, isNaN, isNull, isNullOrUndefined, isNumber, isObject, isPrimitive, isString, isSymbol, isUndefined, mergeDeep, range, rgbToHex, useId };
